### PR TITLE
Support downloading a subset of location tree

### PR DESCRIPTION
**Pull request by orangejenny**
##### SUMMARY
https://dimagi-dev.atlassian.net/browse/ICDS-1123

Patterned after @sravfeyn 's similar feature for mobile workers.

##### FEATURE FLAG
New flag: "Ability to filter location download to include only a specified location and its descendants." @dimagi/product I made this a custom flag, but I could imagine this being useful to non-ICDS projects with a big location hierarchy. Let me know if you agree, and/or have any UI feedback that would make it more useful to other projects.

##### PRODUCT DESCRIPTION
Instead of downloading the entire location tree, users can select a root location and then download only that location and its descendants.

Screen Shot 2020-01-16 at 10 51 46 AM

### CODE REVIEW

File: corehq/apps/locations/util.py [hunk 1]
```diff
@@ -211,6 +214,11 @@ def _write_locations(self, writer, location_type):
         include_consumption = self.include_consumption_flag and location_type.name not in self.administrative_types
         if self.headers_only:
             query = SQLLocation.objects.none()
+        elif self.root_location_id:
+            root_location = SQLLocation.objects.get(location_id=self.root_location_id)
+            query = SQLLocation.active_objects.get_descendants(
+                Q(domain=self.domain, id=root_location.id)
+            ).filter(location_type=location_type)
```
Comment (esoergel) [inline corehq/apps/locations/util.py]: nit: you can also write this expression as
```python
query = root_location.get_descendants().filter(location_type=location_type)
```

File: corehq/apps/locations/views.py [hunk 2]
```diff
@@ -1014,15 +1035,21 @@ def dispatch(self, request, *args, **kwargs):
 
     def get(self, request, *args, **kwargs):
         context = super(DownloadLocationStatusView, self).main_context
+        if toggles.FILTERED_LOCATION_DOWNLOAD.enabled(self.domain):
+            next_url = reverse(FilteredLocationDownload.urlname, args=[self.domain])
+            next_url_text = _("Go back to organization download")
```
Comment (esoergel) [inline corehq/apps/locations/views.py]: I thought this replaced the regular download page?

File: corehq/apps/locations/views.py [hunk 3]
```diff
@@ -222,6 +227,23 @@ def to_json(loc):
             return [to_json(user.get_sql_location(self.domain))]
 
 
+@location_safe
+class FilteredLocationDownload(BaseLocationView):
+    urlname = 'filter_and_download_locations'
+    page_title = ugettext_noop('Filter and Download Locations')
+    template_name = 'locations/filter_and_download.html'
+
+    @method_decorator(require_can_edit_or_view_locations)
```
Comment (esoergel) [inline corehq/apps/locations/views.py]: nit: you can stick this decorator on the class with the arg `name='dispatch'` and not need to override dispatch here.

File: corehq/apps/locations/util.py [hunk 4]
```diff
@@ -211,6 +214,11 @@ def _write_locations(self, writer, location_type):
         include_consumption = self.include_consumption_flag and location_type.name not in self.administrative_types
         if self.headers_only:
             query = SQLLocation.objects.none()
+        elif self.root_location_id:
```
Comment (esoergel) [inline corehq/apps/locations/util.py]: I think this stuff should be abstracted out to a `base_queryset` kind of thing.  There's an `_increment_progress` method that uses the total count of locations on the domain as the total for a progress bar, while this change will of course make that number smaller.

**Review by esoergel — COMMENTED** [review]
Very nice!

This actually opens up the door for another much-desired feature: location export for restricted users.  Right now restricted users can only download [an empty template file](https://github.com/dimagi/commcare-hq/blob/26012aff5ebeddebc2930a16993995217279e48e/corehq/apps/locations/views.py#L978-L979) to populate with new locations.  With this change (and a few minor changes to accommodate it) they should be able to download a subtree if it's entirely accessible to them.  That would make it possible for them to perform updates via the uploader, not just additions.  Something to think about.

File: corehq/apps/locations/util.py (hunk 4 above)
Comment (orangejenny) [reply corehq/apps/locations/util.py]: 292a5968ec3

File: corehq/apps/locations/util.py [hunk 5]
```diff
@@ -117,26 +118,39 @@ def get_location_data_model(domain):
...
         self._location_count = None
         self._locations_exported = 0
 
+        if headers_only:
+            self.base_query = SQLLocation.objects
+        elif root_location_id:
+            root_location = SQLLocation.objects.get(location_id=root_location_id)
+            self.base_query = SQLLocation.active_objects.get_descendants(
+                Q(domain=self.domain, id=root_location.id)
+            ).filter(location_type=location_type)
```
[the remark this answers is not in the document]
Comment (orangejenny) [reply corehq/apps/locations/util.py]: I apparently tested this change locally without restarting celery. Thanks, stickler.

**Comment by esoergel** [comment]
> The additional work would be to integrate these changes into the bulk upload page and then make the dropdown display only the locations a user has access to, does that sound right?

Exactly, and of course validate server side that the user has access to the selected location (and that they selected one)

File: corehq/apps/locations/util.py [hunk 6]
```diff
@@ -117,26 +118,38 @@ def get_location_data_model(domain):
...
-        self.headers_only = headers_only
         self.data_model = get_location_data_model(domain)
         self.administrative_types = {}
         self.location_types = LocationType.objects.by_domain(domain)
         self.async_task = async_task
         self._location_count = None
         self._locations_exported = 0
 
+        if headers_only:
+            self.base_query = SQLLocation.objects
```
Comment (esoergel) [inline corehq/apps/locations/util.py]:
```suggestion
            self.base_query = SQLLocation.objects.none()
```
(this was `.none()` before the move)

**Review by esoergel — CHANGES_REQUESTED** [review]
One blocking comment that I think was a typo, otherwise :shipit:

**Comment by orangejenny** [comment]
@esoergel thank you, fixed

**Review by esoergel — APPROVED** [review]

**Comment by AndreaMKing5** [comment]
Ooh @orangejenny I like this!  Not ready to pull the trigger to make it a Solutions flag but @esoergel will be helpful to keep in mind as we start the release management conversations...

**Pull request merged**